Add "Optimizing Build Performance" section to the Cargo book

Kobzol opened the pull request:
### What does this PR try to resolve?

This PR introduces a new section into the Cargo book, which teaches Rust users how to optimize the build performance of crates. The motivation comes from the results of the Compiler performance survey, where people expressed interest in having an official guide for improving compilation times.

This has been discussed on [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/246057-t-cargo/topic/Build.20performance.20section.20in.20the.20Cargo.20book/with/537603929) briefly.

### How to test and review this PR?

This is not expected to be complete to avoid a very large PR that gets blocked on the details over every part.  Instead this will be merged after beta branch with the expectation that it will be sufficiently filled out by the next beta branch.  Initially, we'll be focusing on content from https://corrode.dev/blog/tips-for-faster-rust-compile-times/.

[Rendered](https://github.com/Kobzol/cargo/blob/performance-guide/src/doc/src/guide/build-performance.md)

nnethercote commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,56 @@
+
+This guide focuses on the first approach.
+
+Below, you can find several methods that can be used to optimize build performance. It is important to nore that their effect varies a lot based on the compiled crate, and in some cases they can actually make compilation slower. You should always measure build performance on your crate(s) to determine if a given method described here is effective for your crate.
```
```suggestion
Below, you can find several methods that can be used to optimize build performance. It is important to note that their effect varies a lot based on the compiled crate, and in some cases they can actually make compilation slower. You should always measure build performance on your crate(s) to determine if a given method described here is effective for your crate.
```

epage commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,56 @@
+
+Note that some of these approaches currently require using the nightly toolchain.
+
+## Reduce amount of generated debug information
```
For a section like this, we've been talking about making changes.  Unsure if there is an issue (and I'm generally cautious about linking to issues in documentation) but I feel like doing so could help gather more insight (e.g. the iteration vs debugging trade off) and give people a sense of "things are moving in a positive direction"

weihanglo commented on src/doc/src/guide/build-performance.md:
````diff
@@ -0,0 +1,56 @@
+codegen-backend = "cranelift"
+```
+
+Since this is currently an unstable option, you will also need to either pass `-Z codegen-backend` to Cargo, or enable this unstable option in the `.cargo/config.toml` file. You can find more information about the unstable `codegen-backend` profile option [here](../reference/unstable.md#codegen-backend).
````
Instead of here, should we do this?

```suggestion
Since this is currently an unstable option, you will also need to either pass `-Z codegen-backend` to Cargo, or enable this unstable option in the `.cargo/config.toml` file. You can find more information in the unstable [`codegen-backend` profile option](../reference/unstable.md#codegen-backend) section.
```

weihanglo commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,56 @@
+
+Note that some of these approaches currently require using the nightly toolchain.
+
+## Reduce amount of generated debug information
```
> I wrote up an approach that applies those ideas to this section at Kobzol#1.

I like the writing style listing tradeoff.

And agree it is overlapped with other reference-like doc.

weihanglo reviewed (commented):
Thanks for writing this draft!

Kobzol commented:
Interesting, your suggestions are more like a "directive" - i.e. "we recommend you to do exactly this to get these benefits", where I tried to talk more broadly about the context of what is being changed. I can't really say what is better! But I do get the concern about duplicating content described elsewhere. I integrated your PR with some additions.

Maybe we should spend some time explaining what's the different between configuring something in `Cargo.toml` vs `.cargo/config.toml`? The maintainers/contributors distinction is IMO not clear enough (unless it's described elsewhere in the book).

I added "✅ Smaller disk usage of the target directory" to the trade-off section for the debuginfo config. I'm not sure if it "belongs" in this guide, but it was something mentioned by a lot of people as a sort of a blocker in the compiler performance survey. Even though it doesn't sound immediately relevant to build performance, it can actually affect it a lot. I know from my experience (with a very limited disk space), that the large size of `target` directories forces me to clean them often, which in turn requires me to recompile stuff more often. Not to mention that on Windows, the I/O cost of writing more data to disk is probably also non-negligible.
Another trade-off is ofc "smaller binary size", but that is IMO completely nonrelevant to build performance, except e.g. for embedded use-cases where you flash the binary to a remote device.

Btw, maybe we should also add a separate section for "Optimizing Runtime Performance" later? :)

epage commented:
> Interesting, your suggestions are more like a "directive" - i.e. "we recommend you to do exactly this to get these benefits", where I tried to talk more broadly about the context of what is being changed. I can't really say what is better! But I do get the concern about duplicating content described elsewhere. I integrated your PR with some additions.

What I would like to convey is "here is a starting point that will likely work for most people" and then give context to help decide whether its worth applying at all or how to tweak it.

My concern with first principles is too many people won't care and just want something to try and we'll lose them with the discussion.

> Maybe we should spend some time explaining what's the difference between configuring something in Cargo.toml vs .cargo/config.toml? The maintainers/contributors distinction is IMO not clear enough (unless it's described elsewhere in the book).

Its nor covered elsewhere.  If you can find a good wait to integrate a break down on on when to use any of the options (manifest, repo config, user config), then that'd be great!

> I added "✅ Smaller disk usage of the target directory" to the trade-off section for the debuginfo config.

Performance isn't just speed, so I'm fine with it so long as our calling it out is proportionate to the effect.

> Another trade-off is ofc "smaller binary size", but that is IMO completely nonrelevant to build performance, except e.g. for embedded use-cases where you flash the binary to a remote device.

We're recommending this for the `dev` profile, so binary size isn't relevant imo. The release profile already has `debug = false`.  This is also why I removed a discussion on profiling as that is generally only relevant for `release`.

epage commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,71 @@
+Note that the Cranelift backend might not support all features used by your crate. It is also available only for a limited set of targets.
+
+
+## Where to apply configuration changes
```
This feels weird to be at the same header level as the rest.  Unsure what we typically do

Kobzol commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,71 @@
+# Optimizing Build Performance
+
+Cargo uses configuration defaults that try to balance various aspects, including debuggability, runtime performance, build performance, binary size and others. Because of that, build performance is sometimes traded off for other benefits which may not be as important for your circumstances. This guide will step you through changes you can make to improve build performance.
```
I see. My original thought was that this page should mostly talk about Cargo options, but if you want to expand that to go "all the way", and also give hints about splitting crates, reducing dependency counts, reducing usage of generics/macros, then it of course makes sense.

In the compiler performance survey, we got quite a lot of open responses that said that people don't want to go change their code to optimize build performance at all, which is understandable (although not very realistic with the current state of things).

epage replied:
I figure it would be good to have things in one place vs having cargo optimizations here and code changes elsewhere.  We similarly have a SemVer compatibility page that covers Cargo and Rust.

> In the compiler performance survey, we got quite a lot of open responses that said that people don't want to go change their code to optimize build performance at all, which is understandable (although not very realistic with the current state of things).

They can always skip those sections :)

epage commented on src/doc/src/guide/build-performance.md:
I was considering bringing up dividing this into two sections
- Root causing build times (`--timings`, compiler passes, macro lines, `cargo llvm-lines`)
- Optimizing builds

I was just split on that vs framing those in terms of the current organization.  For example, we could have a section on reducing the impact of generics and talk through using `cargo llvm-lines` to find large, repeated generic functions and discuss how to improve that.

Kobzol commented:
I also thought that merging this after beta branches might be a reasonable way of giving us enough time to build the content incrementally before the next stable release.

I split the text into two parts (Cargo configuration and source code organization) and rewrote the Cranelift section in the "trade-off format".

epage commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,76 @@
+- ❌ Only available for [certain targets](https://github.com/rust-lang/rustc_codegen_cranelift?tab=readme-ov-file#platform-support)
+- ❌ Might not support all Rust features (e.g. unwinding)
+
+## Source code organization
+TODO
```
Even if we leave merge this after beta branch, I'd rather not be leaving this TODO section in

epage commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,76 @@
+This will change the [`dev` profile](../reference/profiles.md#dev) to use the [Cranelift codegen backend](https://github.com/rust-lang/rustc_codegen_cranelift) for generating machine code, instead of the default LLVM backend. The Cranelift backend should generate code faster than LLVM, which should result in improved build performance.
+
+Trade-offs:
+- ✅ Faster build times
+- ❌ Worse runtime performance of the generated code
```
If I'm understanding correctly, being more specific, this is
- This always speeds up builds of proc macros and build scripts
- For everything else, this only helps with `cargo build` and not `cargo check`
- `cargo build` will mostly be done through `cargo test` and this could make `cargo test` times worse than the savings in `cargo build`

Any way we could provide better nuance here?  Saying "faster build times" may sound more broad than it is.  And saying worse runtime performance, people may not pick up on the implications.

Kobzol commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,76 @@
+
+### Use an alternative codegen backend
+
+> **This requires nightly/unstable features**
```
It is, but I thought that the nightly/unstable aspect of it is the most important thing to consider, that is also why I have been accenting it so much in the guide. IMO nightly will be a show-stopper for many people and listing it just as another trade-off feels wrong.

epage commented on src/doc/src/guide/build-performance.md:
```diff
@@ -0,0 +1,76 @@
+# Optimizing Build Performance
+
+This guide will step you through Cargo configuration options and source code organization patterns that can help improve build performance, by prioritizing it over other aspects which may not be as important for your circumstances.
```
```suggestion
Cargo configuration options and source code organization patterns can help improve build performance, by prioritizing it over other aspects which may not be as important for your circumstances.
```

epage reviewed (approved):
Thanks for the work!  The plan is to merge this after beta branch (the 12th) and continue iterating from there with the hope we have enough content by the next beta branch to justify this section
